tests/unit/s2n_random_test.c

```diff
@@ -894,6 +897,27 @@ int main(int argc, char **argv)
 {
     BEGIN_TEST_NO_INIT();
 
+    /* Feature probe */
+    {
+#if defined(S2N_LIBCRYPTO_SUPPORTS_ENGINE) && defined(OPENSSL_NO_ENGINE)
+        FAIL_MSG("Engine feature probe failed.")
+#endif
+    };
+
+    /* s2n_supports_custom_rand */
+    {
+        if (s2n_supports_custom_rand()) {
+            EXPECT_FALSE(s2n_libcrypto_is_boringssl());
+            EXPECT_FALSE(s2n_libcrypto_is_libressl());
+            EXPECT_FALSE(s2n_libcrypto_is_awslc());
+            EXPECT_FALSE(s2n_is_in_fips_mode());
+        }
+
+#if !defined(S2N_LIBCRYPTO_SUPPORTS_ENGINE)
+        EXPECT_FALSE(s2n_supports_custom_rand());
+#endif
```

Comment:
Does this prove that your feature probe works? What happens if you add a typo to your feature probe?

You want to make sure you have positive and negative tests that run regardless of S2N_LIBCRYPTO_SUPPORTS_ENGINE. So something like:
```
if awslc:
    assert not S2N_LIBCRYPTO_SUPPORTS_ENGINE
if a specific known good version of openssl:
   assert S2N_LIBCRYPTO_SUPPORTS_ENGINE
```

Reply:
The awslc case is covered with `s2n_supports_custom_rand()`, but maybe it makes sense to add a separate assertion since the definition of `s2n_supports_custom_rand` could change?

I had trouble with the positive case. Let me take another swing at it.

Reply:
> The awslc case is covered with s2n_supports_custom_rand()

Let's pretend S2N_LIBCRYPTO_SUPPORTS_ENGINE is wrong, and is "true" for awslc. That define on line 916 is not true, so your test doesn't get compiled, so the test does not detect the mistake. And if S2N_LIBCRYPTO_SUPPORTS_ENGINE is false (correctly or incorrectly), s2n_supports_custom_rand is always false, because s2n_supports_custom_rand depends on S2N_LIBCRYPTO_SUPPORTS_ENGINE. That does not test your feature probe.

Reply:
[Here](https://github.com/aws/s2n-tls/pull/4878/commits/074e2495c4b57360b60f5fa9c8f68512f5301c7f) is my attempt at testing this.

Can you please check my use of `OPENSSL_NO_ENGINE`. Despite official documentation, I had to use it to detect the positive case.